File: src/array_buffer/test_cases.zig
const std = @import("std");
const t = @import("tokota");

const Dba = std.heap.DebugAllocator(.{});
var dba = Dba{};

pub const InstanceData = struct {
    pub fn init(env: t.Env, exports: t.Val) !t.Val {
        try env.instanceDataSet(&InstanceData{}, .with(deinit));
        return exports;
    }

    fn deinit(_: *const InstanceData, _: t.Env) !void {
        return switch (dba.deinit()) {
            .ok => {},
            .leak => t.panic("Memory leak detected", null),
        };
    }
};

pub const ArrayBuffers = struct {
    pub fn create(call: t.Call, input: t.TypedArray(.u8)) !t.ArrayBuffer {
        return call.env.arrayBufferFrom(input.data);
    }

    const OwnedBuffer = struct {
        comptime js_tag: t.Object.Tag = .{
            .lower = 0xab6927a5f2f84a4f,
            .upper = 0x986239b4d2d02641,
        },
        buf: [16]u8,
    };

    pub fn createZigOwned(call: t.Call) !t.ArrayBuffer {
        const allo = dba.allocator();
        const native_data = try allo.create(OwnedBuffer);
        const native_buf = native_data.buf[0..4];

        @memcpy(native_buf, "\xca\xfe\xf0\x0d");

        const buffer = try call.env.arrayBufferOwned(
            native_buf,
            .withHinted(native_data, deinitZigOwned),
        );
        _ = try buffer.ptr.object(call.env).wrap(native_data, .none);

        return buffer;
    }

    pub fn assertZigOwned(
        call: t.Call,
        buf_js: t.ArrayBuffer,
        expected: t.TypedArray(.u8),
    ) !void {
        const buf_zig = try buf_js.ptr
            .object(call.env)
            .unwrap(*OwnedBuffer) orelse return error.InvalidBufferTypeTag;

        var err_buf: [128]u8 = undefined;

        if (&buf_zig.buf != buf_js.data.ptr) return call.env.throwErr(.{
            .code = "AssertionFailed",
            .msg = try std.fmt.bufPrintZ(&err_buf,
                \\
                \\Mismatched Buffer data pointer.
                \\Expected: {*}
                \\Actual:   {*}
            , .{
                &buf_zig.buf,
                buf_js.data.ptr,
            }),
        });

        if (!std.mem.eql(
            u8,
            buf_zig.buf[0..expected.data.len],
            expected.data,
        )) return call.env.throwErr(.{
            .code = "AssertionFailed",
            .msg = try std.fmt.bufPrintZ(&err_buf,
                \\
                \\Mismatched Buffer data.
                \\Expected: {x:0>2}
                \\Actual:   {x:0>2}
            , .{
                expected.data,
                buf_zig.buf[0..expected.data.len],
            }),
        });
    }

    fn deinitZigOwned(data_js: [*]u8, data_zig: *OwnedBuffer, _: t.Env) !void {
        const allo = dba.allocator();
        std.debug.assert(data_js == &data_zig.buf);
        allo.destroy(data_zig);
    }

    pub fn detach(call: t.Call, array_buffer: t.ArrayBuffer) !void {
        if (try array_buffer.isDetached()) return call.env.throwErr(.{
            .code = "AssertionFailed",
            .msg = "Expected non-detached ArrayBuffer as first argument",
        });

        try array_buffer.detach();
        if (!try array_buffer.isDetached()) return call.env.throwErr(.{
            .code = "AssertionFailed",
            .msg = "ArrayBuffer failed to detach",
        });
    }

    pub fn assertDetached(call: t.Call, array_buffer: t.ArrayBuffer) !void {
        if (!try array_buffer.isDetached()) return call.env.throwErr(.{
            .code = "AssertionFailed",
            .msg = "Expected detached ArrayBuffer",
        });
    }

    pub fn isArrayBuffer(call: t.Call, val: t.Val) !bool {
        return val.isArrayBuffer(call.env);
    }
};

pub const Buffers = struct {
    pub fn assertBuffer(call: t.Call, buffer: t.Buffer) !t.Buffer {
        if (!std.mem.eql(u8, buffer.data, "\xca\xfe\xf0\x0d")) {
            var buf: [64]u8 = undefined;

            return call.env.throwErr(.{
                .code = "AssertionFailed",
                .msg = try std.fmt.bufPrintZ(&buf,
                    \\
                    \\Expected: {{ ca, fe, f0, 0d }}
                    \\Actual: {x:0>2}
                , .{buffer.data}),
            });
        }

        return buffer;
    }

    pub fn create(call: t.Call, input: t.TypedArray(.u8)) !t.Buffer {
        const buffer = try call.env.buffer(input.data.len);
        @memcpy(buffer.data, input.data);

        return buffer;
    }

    const OwnedBuffer = struct {
        comptime js_tag: t.Object.Tag = .{
            .lower = 0x5c6bc0aaf16a4016,
            .upper = 0x80eb344d5e91b61b,
        },

        buf: [16]u8,
    };

    pub fn createZigOwned(call: t.Call) !t.Buffer {
        const allo = dba.allocator();
        const native_data = try allo.create(OwnedBuffer);
        const native_buf = native_data.buf[0..4];

        @memcpy(native_buf, "\xca\xfe\xf0\x0d");

        const buffer = try call.env.bufferOwned(
            native_buf,
            .withHinted(native_data, deinitZigOwned),
        );
        _ = try buffer.ptr.object(call.env).wrap(native_data, .none);

        return buffer;
    }

    pub fn assertZigOwned(
        call: t.Call,
        buf_js: t.Buffer,
        expected: t.TypedArray(.u8),
    ) !void {
        const buf_zig = try buf_js.ptr
            .object(call.env)
            .unwrap(*OwnedBuffer) orelse return error.InvalidBufferTypeTag;

        var err_buf: [128]u8 = undefined;

        if (&buf_zig.buf != buf_js.data.ptr) return call.env.throwErr(.{
            .code = "AssertionFailed",
            .msg = try std.fmt.bufPrintZ(&err_buf,
                \\
                \\Mismatched Buffer data pointer.
                \\Expected: {*}
                \\Actual:   {*}
            , .{
                &buf_zig.buf,
                buf_js.data.ptr,
            }),
        });

        if (!std.mem.eql(
            u8,
            buf_zig.buf[0..expected.data.len],
            expected.data,
        )) return call.env.throwErr(.{
            .code = "AssertionFailed",
            .msg = try std.fmt.bufPrintZ(&err_buf,
                \\
                \\Mismatched Buffer data.
                \\Expected: {x:0>2}
                \\Actual:   {x:0>2}
            , .{
                expected.data,
                buf_zig.buf[0..expected.data.len],
            }),
        });
    }

    fn deinitZigOwned(data_js: [*]u8, data_zig: *OwnedBuffer, _: t.Env) !void {
        const allo = dba.allocator();
        std.debug.assert(data_js == &data_zig.buf);
        allo.destroy(data_zig);
    }

    pub fn fromCopy(call: t.Call) !t.Buffer {
        return call.env.bufferFrom("\xca\xfe\xf0\x0d");
    }
};

pub const DataViews = struct {
    pub fn generate(call: t.Call) !t.DataView {
        const array_buffer = try call.env.arrayBuffer(20);

        for (0..array_buffer.data.len) |i| {
            array_buffer.data[i] = @intCast(i);
        }

        return array_buffer.dataView(0, 10);
    }

    pub fn getHttpBody(array_buffer: t.ArrayBuffer) !t.DataView {
        const end_of_header = "\r\n\r\n";
        const idx_eoh = std.mem.lastIndexOf(
            u8,
            array_buffer.data,
            end_of_header,
        ).?;

        const idx_start_of_body = idx_eoh + end_of_header.len;
        const body_len = std.mem.indexOf(
            u8,
            array_buffer.data[idx_start_of_body..],
            "\r\n",
        ).?;

        return array_buffer.dataView(idx_start_of_body, body_len);
    }

    pub fn backingArrayBuffer(call: t.Call, dataview: t.DataView) !t.Val {
        const value = std.mem.readInt(u16, dataview.data[0..2], .big);

        var buf: [64]u8 = undefined;
        if (value != 0x0405) return call.env.throwErr(.{
            .code = "AssertionFailed",
            .msg = try std.fmt.bufPrintZ(&buf,
                \\
                \\Expected: 0x0405
                \\Actual: 0x{x:0>4}
            , .{value}),
        });

        return dataview.buffer;
    }
};

pub const TypedArrays = struct {
    pub fn generateI8(call: t.Call) !t.TypedArray(.i8) {
        const array_buffer = try call.env.arrayBuffer(8);
        const typed_array = try array_buffer.typedArray(.i8, 2, 4);

        typed_array.data[0..4].* = .{ -0x0a, 0x0b, -0x0c, 0x0d };

        return typed_array;
    }

    pub fn generateI8Copy(call: t.Call) !t.TypedArray(.i8) {
        return call.env.typedArrayFrom(&[_]i8{ -0x0a, 0x0b, -0x0c, 0x0d });
    }

    pub fn generateU8(call: t.Call) !t.TypedArray(.u8) {
        const array_buffer = try call.env.arrayBuffer(8);
        const typed_array = try array_buffer.typedArray(.u8, 2, 4);

        typed_array.data[0..4].* = "\xca\xfe\xf0\x0d".*;

        return typed_array;
    }

    pub fn generateU8Copy(call: t.Call) !t.TypedArray(.u8) {
        return call.env.typedArrayFrom("\xca\xfe\xf0\x0d");
    }

    pub fn generateU8Clamped(call: t.Call) !t.TypedArray(.u8c) {
        const array_buffer = try call.env.arrayBuffer(8);
        const typed_array = try array_buffer.typedArray(.u8c, 2, 4);

        typed_array.data[0..4].* = "\xca\xfe\xf0\x0d".*;

        return typed_array;
    }

    pub fn generateI16(call: t.Call) !t.TypedArray(.i16) {
        const array_buffer = try call.env.arrayBuffer(16);
        const typed_array = try array_buffer.typedArray(.i16, 2, 4);

        typed_array.data[0..4].* = .{ -0x0abc, 0x0def, -0x0123, 0x0456 };

        return typed_array;
    }

    pub fn generateI16Copy(call: t.Call) !t.TypedArray(.i16) {
        return call.env.typedArrayFrom(
            &[_]i16{ -0x0abc, 0x0def, -0x0123, 0x0456 },
        );
    }

    pub fn generateU16(call: t.Call) !t.TypedArray(.u16) {
        const array_buffer = try call.env.arrayBuffer(16);
        const typed_array = try array_buffer.typedArray(.u16, 2, 4);

        typed_array.data[0..4].* = .{ 0xcafe, 0xf00d, 0xc001, 0xface };

        return typed_array;
    }

    pub fn generateU16Copy(call: t.Call) !t.TypedArray(.u16) {
        return call.env.typedArrayFrom(
            &[_]u16{ 0xcafe, 0xf00d, 0xc001, 0xface },
        );
    }

    pub fn generateI32(call: t.Call) !t.TypedArray(.i32) {
        const array_buffer = try call.env.arrayBuffer(32);
        const typed_array = try array_buffer.typedArray(.i32, 4, 2);

        typed_array.data[0..2].* = .{ -0x0abc0def, 0x01230456 };

        return typed_array;
    }

    pub fn generateI32Copy(call: t.Call) !t.TypedArray(.i32) {
        return call.env.typedArrayFrom(&[_]i32{ -0x0abc0def, 0x01230456 });
    }

    pub fn generateU32(call: t.Call) !t.TypedArray(.u32) {
        const array_buffer = try call.env.arrayBuffer(32);
        const typed_array = try array_buffer.typedArray(.u32, 4, 2);

        typed_array.data[0..2].* = .{ 0xcafef00d, 0xc001face };

        return typed_array;
    }

    pub fn generateU32Copy(call: t.Call) !t.TypedArray(.u32) {
        return call.env.typedArrayFrom(&[_]u32{ 0xcafef00d, 0xc001face });
    }

    pub fn generateBigI64(call: t.Call) !t.TypedArray(.i64) {
        const array_buffer = try call.env.arrayBuffer(64);
        const typed_array = try array_buffer.typedArray(.i64, 8, 2);

        typed_array.data[0..2].* = .{ -0x0abc0def01230456, 0x012304560abc0def };

        return typed_array;
    }

    pub fn generateBigI64Copy(call: t.Call) !t.TypedArray(.i64) {
        return call.env.typedArrayFrom(
            &[_]i64{ -0x0abc0def01230456, 0x012304560abc0def },
        );
    }

    pub fn generateBigU64(call: t.Call) !t.TypedArray(.u64) {
        const array_buffer = try call.env.arrayBuffer(64);
        const typed_array = try array_buffer.typedArray(.u64, 8, 2);

        typed_array.data[0..2].* = .{ 0xcafef00dc001face, 0xdadad00dfeedcede };

        return typed_array;
    }

    pub fn generateBigU64Copy(call: t.Call) !t.TypedArray(.u64) {
        return call.env.typedArrayFrom(
            &[_]u64{ 0xcafef00dc001face, 0xdadad00dfeedcede },
        );
    }

    pub fn generateF32(call: t.Call) !t.TypedArray(.f32) {
        const array_buffer = try call.env.arrayBuffer(32);
        const typed_array = try array_buffer.typedArray(.f32, 4, 2);

        typed_array.data[0..2].* = .{ -1.23456789, 1.23456789 };

        return typed_array;
    }

    pub fn generateF32Copy(call: t.Call) !t.TypedArray(.f32) {
        return call.env.typedArrayFrom(&[_]f32{ -1.23456789, 1.23456789 });
    }

    pub fn generateF64(call: t.Call) !t.TypedArray(.f64) {
        const array_buffer = try call.env.arrayBuffer(64);
        const typed_array = try array_buffer.typedArray(.f64, 8, 2);

        typed_array.data[0..2].* = .{ -1.23456789, 1.23456789 };

        return typed_array;
    }

    pub fn generateF64Copy(call: t.Call) !t.TypedArray(.f64) {
        return call.env.typedArrayFrom(&[_]f64{ -1.23456789, 1.23456789 });
    }

    pub fn getHttpBody(buf: t.ArrayBuffer) !t.TypedArray(.u8) {
        const end_of_header = "\r\n\r\n";
        const idx_eoh = std.mem.lastIndexOf(u8, buf.data, end_of_header).?;

        const idx_start_of_body = idx_eoh + end_of_header.len;
        const body_len = std.mem.indexOf(
            u8,
            buf.data[idx_start_of_body..],
            "\r\n",
        ).?;

        return buf.typedArray(.u8, idx_start_of_body, body_len);
    }

    pub fn backingArrayBuffer(call: t.Call, arr: t.TypedArray(.u16)) !t.Val {
        const sample_value = std.mem.nativeToBig(u16, arr.data[0]);

        var err_buf: [64]u8 = undefined;
        if (sample_value != 0x0405) return call.env.throwErr(.{
            .code = "AssertionFailed",
            .msg = try std.fmt.bufPrintZ(&err_buf,
                \\
                \\Expected: 0x0405
                \\Actual: 0x{x:0>4}
            , .{sample_value}),
        });

        return arr.buffer;
    }
};

File: src/array_buffer/test_cases.mjs
import assert from "node:assert";
import { Buffer } from "node:buffer";

import { describe, test } from "../testing/runner.mjs";

/**
 * @param {*} addon
 */
export function registerTests(addon) {
  const { ArrayBuffers, Buffers, DataViews, TypedArrays } = addon;

  describe("ArrayBuffer", () => {
    test("create - js-owned", () => {
      /** @type {ArrayBuffer} */
      const buffer = ArrayBuffers.create(Uint8Array.of(0xca, 0xfe, 0xf0, 0x0d));
      assert.equal(new DataView(buffer).getUint32(0, false), 0xcafef00d);
    });

    test("create - zig-owned", () => {
      /** @type {ArrayBuffer} */
      const buffer = ArrayBuffers.createZigOwned();

      const dataView = new DataView(buffer);
      assert.equal(dataView.getUint32(0, false), 0xcafef00d);

      dataView.setUint16(0, 0xface, false);
      ArrayBuffers.assertZigOwned(
        buffer,
        Uint8Array.of(0xfa, 0xce, 0xf0, 0x0d),
      );
    });

    test("detach", () => {
      const arrayBuffer = new ArrayBuffer(16);
      assert(!arrayBuffer.detached);

      ArrayBuffers.detach(arrayBuffer);
      assert(arrayBuffer.detached);
    });

    test("check detached", () => {
      const arrayBuffer = new ArrayBuffer(16);
      arrayBuffer.transfer();
      ArrayBuffers.assertDetached(arrayBuffer);
    });

    test("isArrayBuffer", () => {
      const isArrayBuffer = ArrayBuffers.isArrayBuffer;

      const u8Array = Uint8Array.of(0xca, 0xfe);
      assert.equal(isArrayBuffer(u8Array.buffer), true);

      assert.equal(isArrayBuffer(u8Array), false);
      assert.equal(isArrayBuffer(new DataView(u8Array.buffer)), false);
      assert.equal(isArrayBuffer("nope"), false);
      assert.equal(isArrayBuffer(1996), false);
      assert.equal(isArrayBuffer([0xca, 0xfe]), false);
    });
  });

  describe("TypedArray", () => {
    describe("native-generated", () => {
      test("Int8Array", () => {
        const i8Array = TypedArrays.generateI8();
        assert(i8Array instanceof Int8Array);
        assert.equal(i8Array.buffer.byteLength, 8);
        assert.deepEqual([...i8Array.values()], [-0x0a, 0x0b, -0x0c, 0x0d]);
      });

      test("Int8Array - with Env.typedArrayFrom()", () => {
        const i8Array = TypedArrays.generateI8Copy();
        assert(i8Array instanceof Int8Array);
        assert.equal(i8Array.buffer.byteLength, 4);
        assert.deepEqual([...i8Array.values()], [-0x0a, 0x0b, -0x0c, 0x0d]);
      });

      test("Uint8Array", () => {
        const u8Array = TypedArrays.generateU8();
        assert(u8Array instanceof Uint8Array);
        assert.equal(u8Array.buffer.byteLength, 8);
        assert.deepEqual([...u8Array.values()], [0xca, 0xfe, 0xf0, 0x0d]);
      });

      test("Uint8Array - with Env.typedArrayFrom()", () => {
        const u8Array = TypedArrays.generateU8Copy();
        assert(u8Array instanceof Uint8Array);
        assert.equal(u8Array.buffer.byteLength, 4);
        assert.deepEqual([...u8Array.values()], [0xca, 0xfe, 0xf0, 0x0d]);
      });

      test("Uint8ClampedArray", () => {
        const u8ClampedArray = TypedArrays.generateU8Clamped();
        assert(u8ClampedArray instanceof Uint8ClampedArray);
        assert.equal(u8ClampedArray.buffer.byteLength, 8);
        assert.deepEqual(
          [...u8ClampedArray.values()],
          [0xca, 0xfe, 0xf0, 0x0d],
        );
      });

      test("Int16Array", () => {
        const i16Array = TypedArrays.generateI16();
        assert(i16Array instanceof Int16Array);
        assert.equal(i16Array.buffer.byteLength, 16);
        assert.deepEqual(
          [...i16Array.values()],
          [-0x0abc, 0x0def, -0x0123, 0x0456],
        );
      });

      test("Int16Array - with Env.typedArrayFrom()", () => {
        const i16Array = TypedArrays.generateI16Copy();
        assert(i16Array instanceof Int16Array);
        assert.equal(i16Array.buffer.byteLength, 8);
        assert.deepEqual(
          [...i16Array.values()],
          [-0x0abc, 0x0def, -0x0123, 0x0456],
        );
      });

      test("Uint16Array", () => {
        const u16Array = TypedArrays.generateU16();
        assert(u16Array instanceof Uint16Array);
        assert.equal(u16Array.buffer.byteLength, 16);
        assert.deepEqual(
          [...u16Array.values()],
          [0xcafe, 0xf00d, 0xc001, 0xface],
        );
      });

      test("Uint16Array - with Env.typedArrayFrom()", () => {
        const u16Array = TypedArrays.generateU16Copy();
        assert(u16Array instanceof Uint16Array);
        assert.equal(u16Array.buffer.byteLength, 8);
        assert.deepEqual(
          [...u16Array.values()],
          [0xcafe, 0xf00d, 0xc001, 0xface],
        );
      });

      test("Int32Array", () => {
        const i32Array = TypedArrays.generateI32();
        assert(i32Array instanceof Int32Array);
        assert.equal(i32Array.buffer.byteLength, 32);
        assert.deepEqual([...i32Array.values()], [-0x0abc0def, 0x01230456]);
      });

      test("Int32Array - with Env.typedArrayFrom()", () => {
        const i32Array = TypedArrays.generateI32Copy();
        assert(i32Array instanceof Int32Array);
        assert.equal(i32Array.buffer.byteLength, 8);
        assert.deepEqual([...i32Array.values()], [-0x0abc0def, 0x01230456]);
      });

      test("Uint32Array", () => {
        const u32Array = TypedArrays.generateU32();
        assert(u32Array instanceof Uint32Array);
        assert.equal(u32Array.buffer.byteLength, 32);
        assert.deepEqual([...u32Array.values()], [0xcafef00d, 0xc001face]);
      });

      test("Uint32Array - with Env.typedArrayFrom()", () => {
        const u32Array = TypedArrays.generateU32Copy();
        assert(u32Array instanceof Uint32Array);
        assert.equal(u32Array.buffer.byteLength, 8);
        assert.deepEqual([...u32Array.values()], [0xcafef00d, 0xc001face]);
      });

      test("BigInt64Array", () => {
        const bigI64Array = TypedArrays.generateBigI64();
        assert(bigI64Array instanceof BigInt64Array);
        assert.equal(bigI64Array.buffer.byteLength, 64);
        assert.deepEqual(
          [...bigI64Array.values()],
          [-0x0abc0def01230456n, 0x012304560abc0defn],
        );
      });

      test("BigInt64Array - with Env.typedArrayFrom()", () => {
        const bigI64Array = TypedArrays.generateBigI64Copy();
        assert(bigI64Array instanceof BigInt64Array);
        assert.equal(bigI64Array.buffer.byteLength, 16);
        assert.deepEqual(
          [...bigI64Array.values()],
          [-0x0abc0def01230456n, 0x012304560abc0defn],
        );
      });

      test("BigUint64Array", () => {
        const bigU64Array = TypedArrays.generateBigU64();
        assert(bigU64Array instanceof BigUint64Array);
        assert.equal(bigU64Array.buffer.byteLength, 64);
        assert.deepEqual(
          [...bigU64Array.values()],
          [0xcafef00dc001facen, 0xdadad00dfeedceden],
        );
      });

      test("BigUint64Array - with Env.typedArrayFrom()", () => {
        const bigU64Array = TypedArrays.generateBigU64Copy();
        assert(bigU64Array instanceof BigUint64Array);
        assert.equal(bigU64Array.buffer.byteLength, 16);
        assert.deepEqual(
          [...bigU64Array.values()],
          [0xcafef00dc001facen, 0xdadad00dfeedceden],
        );
      });

      /**
       * @param {number[]} actual
       * @param {number[]} expected
       */
      function assertFloatsEqualish(actual, expected) {
        assert.equal(
          actual.length,
          expected.length,
          "Mismatched float array lengths",
        );

        for (let i = 0; i < actual.length; i += 1) {
          const expectedItem = expected[i];
          assert(expectedItem);

          const actualItem = actual[i];
          assert(actualItem);

          const diff = Math.abs(expectedItem - actualItem);
          assert(
            diff <= 0.00000001,
            `[Index ${i}] Expected ${expectedItem}, got ${actualItem}`,
          );
        }
      }

      test("Float32Array", () => {
        const f32Array = TypedArrays.generateF32();
        assert(f32Array instanceof Float32Array);
        assert.equal(f32Array.buffer.byteLength, 32);
        assertFloatsEqualish([...f32Array.values()], [-1.23456789, 1.23456789]);
      });

      test("Float32Array - with Env.typedArrayFrom()", () => {
        const f32Array = TypedArrays.generateF32Copy();
        assert(f32Array instanceof Float32Array);
        assert.equal(f32Array.buffer.byteLength, 8);
        assertFloatsEqualish([...f32Array.values()], [-1.23456789, 1.23456789]);
      });

      test("Float64Array", () => {
        const f64Array = TypedArrays.generateF64();
        assert(f64Array instanceof Float64Array);
        assert.equal(f64Array.buffer.byteLength, 64);
        assertFloatsEqualish([...f64Array.values()], [-1.23456789, 1.23456789]);
      });

      test("Float64Array - with Env.typedArrayFrom()", () => {
        const f64Array = TypedArrays.generateF64Copy();
        assert(f64Array instanceof Float64Array);
        assert.equal(f64Array.buffer.byteLength, 16);
        assertFloatsEqualish([...f64Array.values()], [-1.23456789, 1.23456789]);
      });
    });

    test("from ArrayBuffer", () => {
      const arrayBuffer = Buffer.from(
        "HTTP/1.1 200 Ok" + "\r\n\r\n" + "\xca\xfe\xf0\x0d" + "\r\n",
        "ascii",
      ).buffer;

      const bodyBytes = TypedArrays.getHttpBody(arrayBuffer);
      assert(bodyBytes instanceof Uint8Array);
      assert.deepEqual([...bodyBytes.values()], [0xca, 0xfe, 0xf0, 0x0d]);
    });

    test("to ArrayBuffer", () => {
      const arrayBuffer = new ArrayBuffer(10);
      new Uint8Array(arrayBuffer).set([
        0x0, 0x1, 0x2, 0x3, 0x4, 0x5, 0x6, 0x7, 0x8, 0x9,
      ]);

      const u16Array = new Uint16Array(arrayBuffer, 4, 2);
      assert.equal(TypedArrays.backingArrayBuffer(u16Array), arrayBuffer);
    });
  });

  describe("DataView", () => {
    test("native-generated", () => {
      const dataView = DataViews.generate();
      assert(dataView instanceof DataView);
      assert.equal(dataView.byteLength, 10);
      assert.equal(dataView.buffer.byteLength, 20);
      assert.equal(dataView.getUint32(2, false), 0x02030405);
    });

    test("from ArrayBuffer", () => {
      const arrayBuffer = Buffer.from(
        "HTTP/1.1 200 Ok" + "\r\n\r\n" + "\xca\xfe\xf0\x0d" + "\r\n",
        "ascii",
      ).buffer;

      const bodyDataView = DataViews.getHttpBody(arrayBuffer);
      assert(bodyDataView instanceof DataView);
      assert.equal(bodyDataView.getUint32(0, false), 0xcafef00d);
    });

    test("to ArrayBuffer", () => {
      const arrayBuffer = Uint8Array.from([
        0x0, 0x1, 0x2, 0x3, 0x4, 0x5, 0x6, 0x7, 0x8, 0x9,
      ]).buffer;

      const dataView = new DataView(arrayBuffer, 4, 2);
      assert.equal(DataViews.backingArrayBuffer(dataView), arrayBuffer);
    });
  });

  describe("Buffer", () => {
    test("receive", () => {
      const buffer = Buffer.from("\xca\xfe\xf0\x0d", "ascii");
      Buffers.assertBuffer(buffer);
    });

    test("create - js-owned", () => {
      /**
       *  @type {import("node:buffer").Buffer}
       */
      const buffer = Buffers.create(Uint8Array.of(0xca, 0xfe, 0xf0, 0x0d));
      assert.equal(buffer.readUint32BE(), 0xcafef00d);
    });

    test("create - zig-owned", () => {
      /**
       *  @type {import("node:buffer").Buffer}
       */
      const buffer = Buffers.createZigOwned(
        Uint8Array.of(0xca, 0xfe, 0xf0, 0x0d),
      );
      assert.equal(buffer.readUint32BE(), 0xcafef00d);

      buffer.writeUint16BE(0xface);
      Buffers.assertZigOwned(buffer, Uint8Array.of(0xfa, 0xce, 0xf0, 0x0d));
    });

    test("create from copy", () => {
      assert.deepEqual(
        Buffers.fromCopy(),
        Buffer.from("\xca\xfe\xf0\x0d", "ascii"),
      );
    });
  });
}
